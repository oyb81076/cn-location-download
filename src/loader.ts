import { load } from "cheerio";
import { decode } from "iconv-lite";
import { get } from "request";
import { resolve } from "url";
export async function loadProvinces(url: string) {
  const $ = await getPage(url);
  const arr: Array<{ name: string, code: number, link: string }> = [];
  $(".provincetr a").each((_, el) => {
    const href = el.attribs.href;
    const code = parseInt(href.split(".")[0], 10) * 100000000;
    const name = $(el).text();
    const link = resolve(url, href);
    arr.push({ name, link, code });
  });
  return arr;
}
export async function loadCities(url: string) {
  return parseBasicPage(url, ".citytr");
}

export function loadCounties(url: string) {
  return parseBasicPage(url, ".countytr");
}
export function loadTowns(url: string) {
  return parseBasicPage(url, ".towntr");
}
export async function loadVillages(url: string) {
  const $ = await getPage(url);
  const arr: Array<{ name: string, code: number }> = [];
  $(".villagetr").each((_, tr) => {
    const codeString = $(tr).children().first().text();
    const code = parseInt(codeString, 10);
    const name = $(tr).children().last().text();
    arr.push({ name, code });
  });
  return arr;
}
export async function parseBasicPage(url: string, selector: string) {
  const $ = await getPage(url);
  const arr: Array<{ name: string, code: number, link?: string }> = [];
  $(selector).each((_, tr) => {
    const $a = $(tr).find("a");
    if ($a.length === 0) {
      const children = tr.childNodes;
      const name = $(children[0]).text();
      const code = parseInt($(children[1]).text(), 10);
      arr.push({ name, code });
    } else {
      const codeString = $(tr).children().first().find("a").text();
      const code = parseInt(codeString, 10);
      const a = $(tr).children().last().find("a");
      const href = a.attr("href");
      const name = a.text();
      const link = resolve(url, href);
      arr.push({ name, link, code });
    }
  });
  return arr;
}
export async function getPage(url: string): Promise<CheerioStatic> {
  return new Promise<CheerioStatic>((r, reject) => {
    get(url, { encoding: null }, (err, res, data) => {
      if (err) {
        reject(err);
      } else if ( res.statusCode < 200 || res.statusCode >= 300 ) {
        const text = decode(data, "gb2312").toString();
        reject(text);
      } else {
        const text = decode(data, "gb2312").toString();
        r(load(text));
      }
    });
  });
}
