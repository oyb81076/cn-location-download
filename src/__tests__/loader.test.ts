import { loadCities, loadCounties, loadProvinces, loadTowns, loadVillages } from "../loader";

test("provinces", async () => {
  const data = await loadProvinces("http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/index.html");
  expect(data).toMatchSnapshot();
});

test("cities", async () => {
  const data = await loadCities("http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/13.html");
  expect(data).toMatchSnapshot();
});
test("county", async () => {
  const data = await loadCounties("http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/1407.html");
  expect(data).toMatchSnapshot();
});
test("town", async () => {
  const data = await loadTowns("http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/140725.html");
  expect(data).toMatchSnapshot();
});
test("vellages", async () => {
  const data = await loadVillages("http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/25/140725200.html");
  expect(data).toMatchSnapshot();
});
