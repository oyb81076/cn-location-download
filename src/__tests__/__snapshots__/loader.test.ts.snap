// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cities 1`] = `
Array [
  Object {
    "code": 130100000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/13/1301.html",
    "name": "石家庄市",
  },
  Object {
    "code": 130200000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/13/1302.html",
    "name": "唐山市",
  },
  Object {
    "code": 130300000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/13/1303.html",
    "name": "秦皇岛市",
  },
  Object {
    "code": 130400000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/13/1304.html",
    "name": "邯郸市",
  },
  Object {
    "code": 130500000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/13/1305.html",
    "name": "邢台市",
  },
  Object {
    "code": 130600000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/13/1306.html",
    "name": "保定市",
  },
  Object {
    "code": 130700000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/13/1307.html",
    "name": "张家口市",
  },
  Object {
    "code": 130800000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/13/1308.html",
    "name": "承德市",
  },
  Object {
    "code": 130900000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/13/1309.html",
    "name": "沧州市",
  },
  Object {
    "code": 131000000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/13/1310.html",
    "name": "廊坊市",
  },
  Object {
    "code": 131100000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/13/1311.html",
    "name": "衡水市",
  },
]
`;

exports[`county 1`] = `
Array [
  Object {
    "code": NaN,
    "name": "140701000000",
  },
  Object {
    "code": 140702000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/140702.html",
    "name": "榆次区",
  },
  Object {
    "code": 140721000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/140721.html",
    "name": "榆社县",
  },
  Object {
    "code": 140722000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/140722.html",
    "name": "左权县",
  },
  Object {
    "code": 140723000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/140723.html",
    "name": "和顺县",
  },
  Object {
    "code": 140724000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/140724.html",
    "name": "昔阳县",
  },
  Object {
    "code": 140725000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/140725.html",
    "name": "寿阳县",
  },
  Object {
    "code": 140726000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/140726.html",
    "name": "太谷县",
  },
  Object {
    "code": 140727000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/140727.html",
    "name": "祁县",
  },
  Object {
    "code": 140728000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/140728.html",
    "name": "平遥县",
  },
  Object {
    "code": 140729000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/140729.html",
    "name": "灵石县",
  },
  Object {
    "code": 140781000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/140781.html",
    "name": "介休市",
  },
]
`;

exports[`provinces 1`] = `
Array [
  Object {
    "code": 1100000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/11.html",
    "name": "北京市",
  },
  Object {
    "code": 1200000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/12.html",
    "name": "天津市",
  },
  Object {
    "code": 1300000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/13.html",
    "name": "河北省",
  },
  Object {
    "code": 1400000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14.html",
    "name": "山西省",
  },
  Object {
    "code": 1500000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/15.html",
    "name": "内蒙古自治区",
  },
  Object {
    "code": 2100000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/21.html",
    "name": "辽宁省",
  },
  Object {
    "code": 2200000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/22.html",
    "name": "吉林省",
  },
  Object {
    "code": 2300000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/23.html",
    "name": "黑龙江省",
  },
  Object {
    "code": 3100000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/31.html",
    "name": "上海市",
  },
  Object {
    "code": 3200000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/32.html",
    "name": "江苏省",
  },
  Object {
    "code": 3300000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/33.html",
    "name": "浙江省",
  },
  Object {
    "code": 3400000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/34.html",
    "name": "安徽省",
  },
  Object {
    "code": 3500000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/35.html",
    "name": "福建省",
  },
  Object {
    "code": 3600000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/36.html",
    "name": "江西省",
  },
  Object {
    "code": 3700000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/37.html",
    "name": "山东省",
  },
  Object {
    "code": 4100000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/41.html",
    "name": "河南省",
  },
  Object {
    "code": 4200000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/42.html",
    "name": "湖北省",
  },
  Object {
    "code": 4300000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/43.html",
    "name": "湖南省",
  },
  Object {
    "code": 4400000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/44.html",
    "name": "广东省",
  },
  Object {
    "code": 4500000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/45.html",
    "name": "广西壮族自治区",
  },
  Object {
    "code": 4600000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/46.html",
    "name": "海南省",
  },
  Object {
    "code": 5000000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/50.html",
    "name": "重庆市",
  },
  Object {
    "code": 5100000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/51.html",
    "name": "四川省",
  },
  Object {
    "code": 5200000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/52.html",
    "name": "贵州省",
  },
  Object {
    "code": 5300000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/53.html",
    "name": "云南省",
  },
  Object {
    "code": 5400000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/54.html",
    "name": "西藏自治区",
  },
  Object {
    "code": 6100000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/61.html",
    "name": "陕西省",
  },
  Object {
    "code": 6200000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/62.html",
    "name": "甘肃省",
  },
  Object {
    "code": 6300000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/63.html",
    "name": "青海省",
  },
  Object {
    "code": 6400000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/64.html",
    "name": "宁夏回族自治区",
  },
  Object {
    "code": 6500000000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/65.html",
    "name": "新疆维吾尔自治区",
  },
]
`;

exports[`town 1`] = `
Array [
  Object {
    "code": 140725100000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/25/140725100.html",
    "name": "朝阳镇",
  },
  Object {
    "code": 140725101000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/25/140725101.html",
    "name": "南燕竹镇",
  },
  Object {
    "code": 140725102000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/25/140725102.html",
    "name": "宗艾镇",
  },
  Object {
    "code": 140725103000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/25/140725103.html",
    "name": "平头镇",
  },
  Object {
    "code": 140725104000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/25/140725104.html",
    "name": "松塔镇",
  },
  Object {
    "code": 140725105000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/25/140725105.html",
    "name": "西洛镇",
  },
  Object {
    "code": 140725106000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/25/140725106.html",
    "name": "尹灵芝镇",
  },
  Object {
    "code": 140725200000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/25/140725200.html",
    "name": "平舒乡",
  },
  Object {
    "code": 140725201000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/25/140725201.html",
    "name": "解愁乡",
  },
  Object {
    "code": 140725202000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/25/140725202.html",
    "name": "温家庄乡",
  },
  Object {
    "code": 140725203000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/25/140725203.html",
    "name": "景尚乡",
  },
  Object {
    "code": 140725204000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/25/140725204.html",
    "name": "上湖乡",
  },
  Object {
    "code": 140725205000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/25/140725205.html",
    "name": "羊头崖乡",
  },
  Object {
    "code": 140725206000,
    "link": "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/14/07/25/140725206.html",
    "name": "马首乡",
  },
]
`;

exports[`vellages 1`] = `
Array [
  Object {
    "code": 140725200200,
    "name": "平舒村委会",
  },
  Object {
    "code": 140725200201,
    "name": "太安村委会",
  },
  Object {
    "code": 140725200202,
    "name": "古城村委会",
  },
  Object {
    "code": 140725200203,
    "name": "西岢村委会",
  },
  Object {
    "code": 140725200204,
    "name": "东岢村委会",
  },
  Object {
    "code": 140725200205,
    "name": "黄岭村委会",
  },
  Object {
    "code": 140725200206,
    "name": "东郭义村委会",
  },
  Object {
    "code": 140725200207,
    "name": "西郭义村委会",
  },
  Object {
    "code": 140725200208,
    "name": "安公村委会",
  },
  Object {
    "code": 140725200209,
    "name": "段王村委会",
  },
  Object {
    "code": 140725200210,
    "name": "上峪村委会",
  },
  Object {
    "code": 140725200211,
    "name": "龙门河村委会",
  },
]
`;
